include/madwf_ml.h

```diff
@@ -0,0 +1,154 @@
+          F<int8_t> w(out, in, args...);
+        } else {
+          errorQuda("QUDA_PRECISION=%d does not enable quarter precision", QUDA_PRECISION);
+        }
+      } break;
+      default: errorQuda("Unsupported precision %d", in.Precision());
+      }
+    }
+  } // namespace madwf_ml
+
+  /**
+    @brief A class for performing the Mobius Accelerated Domain Wall Fermion (MADWF) with the machine learning (ML)
+    version, i.e. the transfer matrix between the vectors with different Ls are _trained_.
+    - As indicated, the parameters, ultimately contained in `device_param`, need to be trained by calling the
+    `train` method.
+    - Once trained, the accelerated operator can be applied with the `apply` method.
+    @param transfer_float the floating-point type we use for the transfer matrix
+    @param trained whether this object has been trained or not
+    @param mu internal copy of the suppression factor
+    @param forward_tmp, backward_tmp persistent ColorSpinorField buffers for reuse
+    @param prec_precondition the underlying preconditioning precision
+    @param host_training_param_cache a static map that holds the cached parameters, so we do not need to load the same
+      set of parameters multiple times from disk
+  */
+  struct MadwfAcc {
+
+    using transfer_float = float;
+
+    static constexpr madwf_ml::transfer_5D_t transfer_t = madwf_ml::transfer_5D_t::Spin;
+
+    using device_container = device_vector<transfer_float>;
+
+    bool trained = false;
+
+  private:
+    device_container device_param;
+
+    MadwfParam param;
+
+    double mu;
+
+    std::unique_ptr<ColorSpinorField> forward_tmp;
+    std::unique_ptr<ColorSpinorField> backward_tmp;
+
+    QudaPrecision prec_precondition;
+
+    static std::unordered_map<std::string, std::vector<transfer_float>> host_training_param_cache; // empty map
+
+    TimeProfile &profile;
+
+    /**
+      @brief Fill a host vector with Gaussian random numbers.
+      @param[in] v the host vector
+    */
+    void fill_random(std::vector<transfer_float> &v);
+
+    /**
+      @brief Return the chi squared of in the machine larning training.
+      @param[in] ref the reference operator
+      @param[in] base the base solver
+      @param[out] out the output vector
+      @param[in] in the input vector
+    */
+    double cost(const DiracMatrix &ref, Solver &base, ColorSpinorField &out, const ColorSpinorField &in);
+
+    /**
+      @brief Save the current parameter to disk
+      @param[in] Ls the original (larger) 5th dimension size
+      @param[in] Ls_base the reduced 5th dimension size
+    */
+    void save_parameter(int Ls, int Ls_base);
+
+    /**
+      @brief Load the parameter from disk
+      @param[in] Ls the original (larger) 5th dimension size
+      @param[in] Ls_base the reduced 5th dimension size
+      @param[in] param_size the expected parameter size
+    */
+    void load_parameter(int Ls, int Ls_base, size_t param_size);
```

Comment:
Why do we need to pass in the `param_size`? Shouldn't the loading routine know exactly how large it should be in advance?

Reply:
https://github.com/lattice/quda/pull/1061/commits/7ef32163893b0cab221e561ff27c28c696576277